data/kernels/basic.cl

```diff
@@ -677,27 +720,24 @@ tonecurve (read_only image2d_t in, write_only image2d_t out, const int width, co
   else if(tc_mode == 3)
   {
     float4 rgb = Lab_to_prophotorgb(pixel);
-    rgb.x = lookup_unbounded(table_0, rgb.x, coeffs_0);
-    rgb.y = lookup_unbounded(table_0, rgb.y, coeffs_0);
-    rgb.z = lookup_unbounded(table_0, rgb.z, coeffs_0);
-    pixel.xyz = prophotorgb_to_Lab(rgb).xyz;
+    float norm = 0;
+    if (rgb_norm != 0)
+    {
+      norm = rgb_norm_vect( rgb, rgb_norm, 0.333333f);
+      norm = lookup_unbounded(table_0, norm, coeffs_0) / norm;
+      rgb.xyz *= norm;
```

Comment:
you got that wrong :-)

```C
norm = rgb_norm_vect( rgb, rgb_norm, 0.333333f); // compute the norm == luminance estimator
const float4 rgb_ratios = rgb /norm; // these ratios are the actual colors, independant from the luminance
norm = lookup_unbounded(table_0, norm, coeffs_0); // compute the curve on the luminance
rgb.xyz = (norm * rgb_ratios).xyz; // restore the colors from the original ratios and the new luminance
```

Reply:
I understand this (keep the true color) with norm calculation. 
But in case of norms from your referred article, the goal is to fix some perceptual artefacts at the toe and shoulder of S curve. So the calculation I’ve proposed may be more appropriate in that case. In any case I’ll make some experiment for those.

Reply:
I understand this (keep the true color) with norm calculation. 
But in case of norms from your referred article, the goal is to fix some perceptual artefacts at the toe and shoulder of S curve. So the calculation I’ve proposed may be more appropriate in that case. In any case I’ll make some experiment for those.